pam/internal/adapter/utils.go

```diff
@@ -65,6 +73,46 @@ func isSSHSession(mTx pam.ModuleTransaction) bool {
 	return isSSHSessionValue
 }
 
+func getPamTty(mTx pam.ModuleTransaction) (*os.File, func(), error) {
+	pamTty, err := mTx.GetItem(pam.Tty)
+	if err != nil {
+		return nil, func() {}, err
+	}
+
+	if pamTty == "" {
+		return nil, func() {}, nil
+	}
+
+	tty, err := os.OpenFile(pamTty, os.O_RDWR, 0600)
+	if err != nil {
+		return nil, func() {}, err
+	}
+	cleanup := func() { tty.Close() }
+
+	// We check the fd could be passed to x/term to decide if we should fallback to stdin
+	if tty.Fd() > math.MaxInt {
+		return nil, cleanup, fmt.Errorf("unexpected large PAM TTY fd: %d", tty.Fd())
```

Comment:
Maybe do the cleanup here? It seems a bit unexpected that the caller needs to do cleanup when an error is returned.